pkg/controller/podautoscaler/horizontal_test.go

```diff
@@ -330,38 +330,43 @@ func (tc *testCase) prepareTestClient(t *testing.T) (*fake.Clientset, *metricsfa
 	})
 
 	fakeClient.AddReactor("update", "horizontalpodautoscalers", func(action core.Action) (handled bool, ret runtime.Object, err error) {
-		tc.Lock()
-		defer tc.Unlock()
-
-		obj := action.(core.UpdateAction).GetObject().(*autoscalingv1.HorizontalPodAutoscaler)
-		assert.Equal(t, namespace, obj.Namespace, "the HPA namespace should be as expected")
-		assert.Equal(t, hpaName, obj.Name, "the HPA name should be as expected")
-		assert.Equal(t, tc.expectedDesiredReplicas, obj.Status.DesiredReplicas, "the desired replica count reported in the object status should be as expected")
-		if tc.verifyCPUCurrent {
-			if assert.NotNil(t, obj.Status.CurrentCPUUtilizationPercentage, "the reported CPU utilization percentage should be non-nil") {
-				assert.Equal(t, tc.CPUCurrent, *obj.Status.CurrentCPUUtilizationPercentage, "the report CPU utilization percentage should be as expected")
+		handled, obj, err := func() (handled bool, ret *autoscalingv1.HorizontalPodAutoscaler, err error) {
```

Comment:
You need to apply this same fix to `legacy_horizontal_test.go` as well.

Looks like we were racing to fix the same race condition!  https://github.com/kubernetes/kubernetes/pull/76026/files

FYI: You can reliably reproduce this deadlock with `export GOMAXPROCS=1` and `go test ./pkg/controller/podautoscaler/... -timeout 60s -count 1`.  This PR fail on the legacy test, but your fix fixes this one!

Reply:
Thanks, I added the legacy test fix as well.

If it's all the same to you, I'd like to keep this PR going just because I merged a similar PR in OpenShift and referenced this PR. Our tooling that coordinates with this repo will be waiting for this PR to come through in a future OpenShift rebase so it can drop that patch.

Reply:
No problem.  Thanks for the additional fix in the legacy tests.  Passes now with GOMAXPROCS=1.